150 - Ids Date Picker Component

andriivarhanov opened the pull request:
**Explain the details for making this change. What existing problem does the pull request solve?**
This PR adds `ids-date-picker` component. Adds `ids-date-picker` to `ids-week-view` calendar toolbar and to `ids-month-view` calendar toolbar.
Variation of the date picker where it's dropdown with month/year picker is not included in the PR.

**Related github/jira issue (required)**:
Closes #150 
Related #513 

**Issues**:
There are cycle imports in `ids-date-picker` and `ids-month-view` components. Date picker component uses month view component what uses date picker component... Altrough it doesn't cause any issues for now, I'm not sure it won't cause any issues going further. I’ve eslint disabled the import for now. I would appreciate any help on what can be done with it

**Steps necessary to review your pull request (required)**:
- pull/build/run
- go to http://localhost:4300/ids-date-picker
- compare to the design https://main-enterprise.demo.design.infor.com/components/datepicker/example-index.html
- go to http://localhost:4300/ids-week-view check date picker added to the calendar toolbar
- go to http://localhost:4300/ids-month-view check date picker added to the calendar toolbar

**Included in this Pull Request**:
- [x] An e2e or functional test for the bug or feature.
~- [ ] A note to the change log.~

danielortiz1982 commented on src/components/ids-date-picker/ids-date-picker.js:
```diff
@@ -0,0 +1,786 @@
+      this.#attachMonthView();
+      this.#popup.visible = true;
+      const { bottom } = this.#triggerButton.getBoundingClientRect();
+      const positionBottom = (bottom + 100) < window.innerHeight;
```
This looks great. Just a quick question, do you think the `positionBottom` can be done with css? Also how would this respond to a viewport change? I am a little curious.

tmcconechy replied:
To add to this i think i did this in dropdown by using a different target to position https://github.com/infor-design/enterprise-wc/blob/main/src/components/ids-dropdown/ids-dropdown.js#L338

andriivarhanov replied:
The idea was to position the popup depends on if it fits to viewport size (edges), something like on the screenshots below, but I think it should be a part of ids-popup functionality since some other components might need it (like ids-time-picker for example).  Simplified it for date picker to be bottom left aligned

![image](https://user-images.githubusercontent.com/5204821/151802208-d80e2728-f156-4924-ab99-38398af7e52c.png)
![image](https://user-images.githubusercontent.com/5204821/151802597-122c9931-78af-47ca-b1ba-918406e8382b.png)
![image](https://user-images.githubusercontent.com/5204821/151802988-bb3d7783-f144-4829-9791-43fe89cdbabd.png)

danielortiz1982 commented on src/components/ids-month-view/ids-month-view.js:
```diff
@@ -186,26 +200,36 @@ class IdsMonthView extends Base {
     if (this.showToday) {
       this.offEvent('click.month-view-today');
-      this.onEvent('click.month-view-today', this.container.querySelector('.month-view-btn-today'), () => {
+      this.onEvent('click.month-view-today', this.container.querySelector('.month-view-btn-today'), (e) => {
```
Same... You can consolidate this into one onEvent call.

danielortiz1982 reviewed (changes requested):
If possible reduce the amount of onEvent were possible by checking for event target.

tmcconechy commented on src/components/ids-date-picker/README.md:
```diff
@@ -0,0 +1,49 @@
+- `size` {'xs'|'sm'|'mm'|'md'|'lg'|'full'} - Size (width) of the field
+- `validate` {'required'|string} - Input validation rules
+- `validation-events` {string} - Input validation events, `change blur` as default
+- `format` {'locale'|string|null} - Input date format, if no set defaults to locale calendar date format. Examples: `yyyy-MM-dd`, `d/M/yyyy`, `dd/MM/yyyy`
```
```suggestion
- `format` {'locale'|string|null} - Input date format, if not set defaults to locale calendar date format. Examples: `yyyy-MM-dd`, `d/M/yyyy`, `dd/MM/yyyy`
```

tmcconechy commented on src/components/ids-date-picker/README.md:
```diff
@@ -0,0 +1,49 @@
+- `month` `{string|number|null}` - Specifies a month for the popup calendar (`ids-month-view` attribute)
+- `day` `{string|number|null}` - Specifies a day for the popup calendar (`ids-month-view` attribute)
+- `year` `{string|number|null}` - Specifies a year for the popup calendar (`ids-month-view` attribute)
+- `first-day-of-week` `{string|number|null}` - Specifies first day of the week for the popup calendar (`ids-month-view` attribute)
```
```suggestion
- `first-day-of-week` `{string|number|null}` - Specifies first day of the week for the popup calendar, if not set the information comes from the locale (`ids-month-view` attribute)
```

tmcconechy commented on src/components/ids-date-picker/ids-date-picker.js:
```diff
@@ -0,0 +1,746 @@
+import IdsDropdown from '../ids-dropdown/ids-dropdown';
+import IdsIcon from '../ids-icon/ids-icon';
+import IdsInput from '../ids-input/ids-input';
+// eslint-disable-next-line import/no-cycle
```
We might want to try to find a way to fix the cycle/not just the warning.. @danielortiz1982 ?

tmcconechy replied:
Yeah i know the concept just thinking we should try to find a way to avoid it in this case if its possible (make a shared util ect)

tmcconechy commented on src/components/ids-date-picker/ids-date-picker.js:
```diff
@@ -0,0 +1,746 @@
+ * @mixes IdsThemeMixin
+ * @mixes IdsLocaleMixin
+ */
+
```
Space

tmcconechy commented on src/components/ids-date-picker/ids-date-picker.js:
```diff
@@ -0,0 +1,746 @@
+ * @mixes IdsKeyboardMixin
+ * @mixes IdsPopupOpenEventsMixin
+ * @mixes IdsThemeMixin
+ * @mixes IdsLocaleMixin
```
Could add parts as well.

tmcconechy commented on test/ids-date-picker/ids-date-picker-func-test.js:
```diff
@@ -0,0 +1,315 @@
+    expect(component.year).toBeNull();
+    expect(component.day).toBeNull();
+  });
+});
```
Just FYI you can also nest the describes something like this

```
describe('IdsDatePicker Tests', () => {
  describe('Using Properties', () => {
    it ('should do something', () => {
      expect(true).toBe(true);
    });
  });
    describe('Using Attributes', () => {
    it ('should do something', () => {
      expect(true).toBe(true);
    });
  });
});
```

tmcconechy reviewed (commented):
This looks great! The only comments i noticed on the functionality is the keyboard.

- down arrow - opens the list but doesnt focus a day
- moving with the arrow keys to select days doesnt work

Full spec here https://github.com/infor-design/enterprise/blob/main/src/components/datepicker/readme.md#keyboard-shortcuts

Some can be added to a TODO tho so maybe some basics.

andriivarhanov commented:
@danielortiz1982 @tmcconechy thanks for your feedback, it's still WIP, but I will address all your comments while completing the task

andriivarhanov commented:
@danielortiz1982 @tmcconechy 
I've made changes according to your comments:
- reduced the amount of onEvent in month view calendar toolbar
- added keyboard events TODOs for ids-month-view and ids-date-picker
- added nested describes to func tests
- added themeable parts
- simplified the popup position
please let me know if I missed anything

tmcconechy reviewed (approved):
This looks good to me. Although one thing im thinking is if we organized the example more like dropdown in terms of whats shown it could be more consistent in the examples.

But this works for now so accepting and other inputs have similar inconsistencies

danielortiz1982 commented:
I think this looks great as well. One thing I did notice is once a date is clicked and selected it automatically apples the date and closes the dropdown. Is this the expected behavior or am I suppose to apply the given section?  One other thing, the month year dropdown does not open. If these are not issues I am good to approve it.

https://www.loom.com/share/2d78e18afa1140919f5196b12d07b9d2

andriivarhanov commented:
@danielortiz1982 
> once a date is clicked and selected it automatically apples

it's the expected behavior, same as the original https://main-enterprise.demo.design.infor.com/components/datepicker/example-index.html

> the month year dropdown does not open

the month year picker is not included in the PR, it’s more focused on the date picker as a form field and to add it to month/week view calendar toolbars

danielortiz1982 commented:
> @danielortiz1982
> 
> > once a date is clicked and selected it automatically apples
> 
> it's the expected behavior, same as the original https://main-enterprise.demo.design.infor.com/components/datepicker/example-index.html
> 
> > the month year dropdown does not open
> 
> the month year picker is not included in the PR, it’s more focused on the date picker as a form field and to add it to month/week view calendar toolbars

Perfect! Thanks for clarifying @andriivarhanov 👍

danielortiz1982 reviewed (approved):
This look great! Good job on this @andriivarhanov 👍
<img width="520" alt="Screen Shot 2022-01-31 at 11 10 59 AM" src="https://user-images.githubusercontent.com/5215954/151830193-ba9f3d02-84df-483c-b8a9-1d2926b3773b.png">

EdwardCoyle reviewed (approved):
Really nice work @andriivarhanov.  My only comment is already in the TODO for adding keyboard shortcuts. 👍
